include/machine_event.hh

```diff
@@ -0,0 +1,45 @@
+#ifndef HSIM_MACHINE_EVENT_INCLUDED
+#define HSIM_MACHINE_EVENT_INCLUDED
+
+#include <list>
+#include <memory>
+#include <utility>
+
+#include "support.hh"
+
+namespace hsim {
+
+struct MemRead {};
+struct MemWrite {};
+struct PreInsn {};
+struct PostInsn {};
+
+class IEventConsumer {
+  public:
+    virtual ~IEventConsumer() = default;
+    virtual void handle(MemRead event, Addr addr) = 0;
+    virtual void handle(MemWrite event, Addr addr, Word value) = 0;
+    virtual void handle(PreInsn event, Word insn) = 0;
+    virtual void handle(PostInsn event, Word insn) = 0;
```

Comment:
I'm a bit confused with interface duplication. (Same functions are defined in IPLugin). Can we somehow avoid it? Maybe we can inherit IPlugin from IEventConsumer?